File: python/run_optimization.py
"""run_optimization.py
Genetic algorithm optimisation of cyst-counting parameters using DEAP.

Reads the Optimization set from data/, runs the GA, and writes results
to results/optimization/.

This is a Python translation of run_optimization.m.
See python/README.md for differences from the MATLAB implementation.
"""

import os
import sys
import random
import time
import warnings
from pathlib import Path

import numpy as np
import pandas as pd
import matplotlib
matplotlib.use('Agg')
import matplotlib.pyplot as plt

from deap import base, creator, tools, algorithms

# Ensure the python/ directory is on the path when called from elsewhere
sys.path.insert(0, str(Path(__file__).parent))
from fitness_function import evaluate_params


# ---- Parameter bounds (must match MATLAB) ---------------------------------
PARAM_NAMES = ['minPeakSepSmall', 'minPeakSepLarge', 'largeCystVolThresh',
               'maxPeakDistMerge', 'thresholdFraction', 'gaussianSigma']
LB = np.array([1,  10,  10000, 1,  0.1, 0.01])
UB = np.array([15, 100, 150000, 15, 3.0, 1.5])

# ---- GA hyperparameters (mirror MATLAB) -----------------------------------
POP_SIZE     = 50
N_GEN        = 200
ELITE_COUNT  = 2
CX_PROB      = 0.5      # probability of crossover per pair
MUT_PROB     = 0.95     # probability of mutation per individual
TOURNSIZE    = 3
STALL_LIMIT  = 5        # generations without improvement before diversity injection
REINIT_FRAC  = 0.5      # fraction of non-elite population reinitialised on stall


def _eval_individual(ind, gt_masks, gt_counts):
    """Evaluate one individual sequentially (no subprocesses – mirrors MATLAB's UseParallel=false)."""
    try:
        if any(not np.isfinite(v) or v <= 0 for v in ind):
            return (1e6,)
        score, _ = evaluate_params(ind, gt_masks, gt_counts, n_workers=1)
        return (score if np.isfinite(score) and score >= 0 else 1e6,)
    except Exception:
        return (1e6,)


def _id_to_filename(sample_id: str) -> str:
    prefix = 'KRATS_' if sample_id[0] == 'K' else 'XRATS_'
    return f'{prefix}{int(sample_id[1:]):03d}.nii.gz'


def _clip(individual):
    """Clip individual to parameter bounds in-place."""
    for i, (lo, hi) in enumerate(zip(LB, UB)):
        individual[i] = float(np.clip(individual[i], lo, hi))
    return individual


def run_optimization():
    repo_root  = Path(__file__).resolve().parent.parent
    data_dir   = repo_root / 'data' / '2_Cyst_counting_pipeline' / 'Optimization set'
    mask_dir   = data_dir / 'Cyst masks'
    output_dir = repo_root / 'results' / 'optimization'
    output_dir.mkdir(parents=True, exist_ok=True)

    # ---- Load ground truth ------------------------------------------------
    xls_file = data_dir / 'O1_count.xlsx'
    if not xls_file.is_file():
        sys.exit(f'Ground-truth file not found:\n  {xls_file}\nPlace the Zenodo data inside data/.')

    df = pd.read_excel(xls_file)
    gt_masks  = []
    gt_counts = []
    for _, row in df.iterrows():
        sid  = str(row.iloc[0])
        cnt  = int(row.iloc[1])
        path = mask_dir / _id_to_filename(sid)
        if not path.is_file():
            sys.exit(f'Mask not found for {sid}:\n  {path}')
        gt_masks.append(str(path))
        gt_counts.append(cnt)

    gt_counts = np.array(gt_counts)

    print('\n' + '='*50)
    print('  GENETIC ALGORITHM OPTIMISATION')
    print('='*50)
    print(f'Training samples : {len(gt_counts)}')
    print(f'Count range      : [{gt_counts.min()} – {gt_counts.max()}]  '
          f'(mean {gt_counts.mean():.1f} ± {gt_counts.std():.1f})\n')

    # ---- DEAP setup -------------------------------------------------------
    # Minimisation problem
    if hasattr(creator, 'FitnessMin'):
        del creator.FitnessMin
    if hasattr(creator, 'Individual'):
        del creator.Individual
    creator.create('FitnessMin', base.Fitness, weights=(-1.0,))
    creator.create('Individual', list, fitness=creator.FitnessMin)

    toolbox = base.Toolbox()

    def random_individual():
        ind = [float(LB[i] + random.random() * (UB[i] - LB[i])) for i in range(6)]
        return creator.Individual(ind)

    toolbox.register('individual', random_individual)
    toolbox.register('population', tools.initRepeat, list, toolbox.individual)

    def eval_individual(ind):
        return _eval_individual(ind, gt_masks, gt_counts)

    toolbox.register('evaluate', eval_individual)
    toolbox.register('select',   tools.selTournament, tournsize=TOURNSIZE)
    # cxBlend is the closest available equivalent to MATLAB's adaptive feasible crossover
    toolbox.register('mate',     tools.cxBlend, alpha=0.5)
    # mutPolynomialBounded is the closest equivalent to MATLAB's adaptive feasible mutation
    toolbox.register('mutate',   tools.mutPolynomialBounded,
                     low=LB.tolist(), up=UB.tolist(),
                     eta=20.0, indpb=1.0 / 6)

    # ---- Initialise population --------------------------------------------
    population = toolbox.population(n=POP_SIZE)
    hof        = tools.HallOfFame(ELITE_COUNT)

    # Evaluate initial population sequentially (mirrors MATLAB's UseParallel=false)
    print('Evaluating initial population...', flush=True)
    for i, ind in enumerate(population):
        ind.fitness.values = eval_individual(ind)
        if (i + 1) % 10 == 0:
            print(f'  {i+1}/{POP_SIZE} individuals evaluated', flush=True)

    hof.update(population)

    # ---- History tracking -------------------------------------------------
    history = {
        'generation': [], 'best_score': [], 'MPE': [], 'MAE': [],
        'STD': [], 'R2': [], 'mean_score': [],
    }

    best_ever  = hof[0].fitness.values[0]
    stall_gen  = 0

    print(f'{"Gen":>6}  {"Best MAPE":>12}  {"MPE":>8}  {"MAE±STD":>14}  {"R²":>8}  {"Stall":>6}')
    print('-' * 70)

    # ---- Main GA loop -----------------------------------------------------
    for gen in range(1, N_GEN + 1):
        # Selection (exclude elite slots)
        elite    = tools.selBest(population, ELITE_COUNT)
        offspring = toolbox.select(population, POP_SIZE - ELITE_COUNT)
        offspring = [toolbox.clone(o) for o in offspring]

        # Crossover
        for child1, child2 in zip(offspring[::2], offspring[1::2]):
            if random.random() < CX_PROB:
                toolbox.mate(child1, child2)
                del child1.fitness.values
                del child2.fitness.values

        # Mutation
        for mutant in offspring:
            if random.random() < MUT_PROB:
                toolbox.mutate(mutant)
                _clip(mutant)
                del mutant.fitness.values

        # Evaluate new individuals sequentially
        invalid = [ind for ind in offspring if not ind.fitness.valid]
        for ind in invalid:
            ind.fitness.values = eval_individual(ind)

        # Combine elite + offspring
        population[:] = list(elite) + offspring
        hof.update(population)

        # Stats
        best_score = hof[0].fitness.values[0]
        valid_scores = [ind.fitness.values[0] for ind in population
                        if np.isfinite(ind.fitness.values[0]) and ind.fitness.values[0] < 1e5]
        mean_score = np.mean(valid_scores) if valid_scores else float('nan')

        _, m = evaluate_params(list(hof[0]), gt_masks, gt_counts, n_workers=1)

        if best_score < best_ever - 1e-6:
            best_ever = best_score
            stall_gen = 0
            print(f'{gen:>6}  {best_score:>12.4f}  {m["MPE"]:>8.2f}  '
                  f'{m["MAE"]:>6.2f}±{m["STD"]:<5.2f}  {m["R_squared"]:>8.4f}')
        else:
            stall_gen += 1
            print(f'{gen:>6}  {best_score:>12.4f}  {m["MPE"]:>8.2f}  '
                  f'{m["MAE"]:>6.2f}±{m["STD"]:<5.2f}  {m["R_squared"]:>8.4f}  {stall_gen:>6}')

        # Diversity injection
        if stall_gen >= STALL_LIMIT:
            print(f'   >>> Diversity injection: reinitialising {int(REINIT_FRAC*100)}% of population <<<')
            n_reinit = round(REINIT_FRAC * (POP_SIZE - ELITE_COUNT))
            reinit_idx = random.sample(range(ELITE_COUNT, POP_SIZE), n_reinit)
            for idx in reinit_idx:
                population[idx] = random_individual()
                population[idx].fitness.values = eval_individual(population[idx])
            stall_gen = 0

        history['generation'].append(gen)
        history['best_score'].append(best_score)
        history['MPE'].append(m['MPE'])
        history['MAE'].append(m['MAE'])
        history['STD'].append(m['STD'])
        history['R2'].append(m['R_squared'])
        history['mean_score'].append(mean_score)

    # ---- Final results ----------------------------------------------------
    best_params = list(hof[0])
    best_score  = hof[0].fitness.values[0]
    _, final_m  = evaluate_params(best_params, gt_masks, gt_counts, n_workers=1)

    print('\n' + '='*60)
    print('FINAL RESULTS')
    print('='*60)
    print('\nOptimised parameters:')
    for name, val in zip(PARAM_NAMES, best_params):
        print(f'  {name:<22} : {val:12.6f}')
    print(f'\nTraining-set metrics (n={len(gt_counts)}):')
    print(f'  MAPE : {best_score:8.4f} %')
    print(f'  MPE  : {final_m["MPE"]:8.2f} %')
    print(f'  MAE  : {final_m["MAE"]:8.2f} ± {final_m["STD"]:.2f} cysts')
    print(f'  RMSE : {final_m["RMSE"]:8.2f} cysts')
    print(f'  R²   : {final_m["R_squared"]:8.4f}')

    # ---- Save outputs -----------------------------------------------------
    np.savetxt(output_dir / 'BestParams.txt', best_params)

    hist_df = pd.DataFrame(history)
    hist_df.to_csv(output_dir / 'FitnessHistory.txt', index=False, sep='\t')

    _plot_convergence(history, output_dir)
    _plot_predictions(final_m, gt_counts, output_dir)
    _write_report(best_params, best_score, final_m, history, output_dir, gt_counts)

    print(f'\nResults written to: {output_dir}\n')


# ---- Plotting helpers -----------------------------------------------------

def _plot_convergence(history, output_dir):
    fig, axes = plt.subplots(2, 3, figsize=(16, 9))

    ax = axes[0, 0]
    ax.plot(history['generation'], history['best_score'], 'b-o', ms=4, lw=2, label='Best')
    ax.plot(history['generation'], history['mean_score'], 'r--', lw=1.5, label='Mean')
    ax.set_xlabel('Generation'); ax.set_ylabel('MAPE (%)'); ax.set_title('Fitness convergence')
    ax.legend(); ax.grid(True)

    ax = axes[0, 1]
    ax.errorbar(history['generation'], history['MAE'], yerr=history['STD'],
                fmt='b-o', ms=4, lw=2, capsize=5)
    ax.set_xlabel('Generation'); ax.set_ylabel('MAE (cysts)'); ax.set_title('MAE ± STD')
    ax.grid(True)

    ax = axes[0, 2]
    ax.plot(history['generation'], history['R2'], 'g-o', ms=4, lw=2)
    ax.set_xlabel('Generation'); ax.set_ylabel('R²'); ax.set_title('R² convergence')
    ax.set_ylim([0, 1]); ax.grid(True)

    ax = axes[1, 0]
    ax.plot(history['generation'], history['MPE'], 'm-o', ms=4, lw=2)
    ax.axhline(0, color='k', ls='--', lw=1.5)
    ax.set_xlabel('Generation'); ax.set_ylabel('MPE (%)'); ax.set_title('Bias convergence')
    ax.grid(True)

    # Remove unused subplots
    axes[1, 1].set_visible(False)
    axes[1, 2].set_visible(False)

    fig.suptitle('GA convergence', fontsize=14, fontweight='bold')
    fig.tight_layout()
    fig.savefig(output_dir / 'GA_convergence.png', dpi=150)
    plt.close(fig)


def _plot_predictions(m, gt_counts, output_dir):
    gt   = m['ground_truth']
    pred = m['predictions']
    lo   = min(gt.min(), pred.min())
    hi   = max(gt.max(), pred.max())
    pad  = (hi - lo) * 0.1
    ax_range = [lo - pad, hi + pad]
    p = np.polyfit(gt, pred, 1)

    fig, ax = plt.subplots(figsize=(7, 6))
    ax.scatter(gt, pred, s=80, alpha=0.8, color='#3366CC', zorder=5)
    ax.plot(ax_range, ax_range, 'k--', lw=2, label='Identity')
    xf = np.linspace(ax_range[0], ax_range[1], 100)
    ax.plot(xf, np.polyval(p, xf), 'r-', lw=2, label='Regression')
    ax.set_xlim(ax_range); ax.set_ylim(ax_range)
    ax.set_xlabel('Ground truth (cysts)', fontweight='bold')
    ax.set_ylabel('Predicted (cysts)',    fontweight='bold')
    ax.set_title('Predictions vs ground truth (training set)', fontweight='bold')
    txt = (f'y = {p[0]:.3f}x + {p[1]:.2f}\n'
           f'MAPE: {m["MAPE"]:.2f}%\nR²: {m["R_squared"]:.4f}\n'
           f'MAE: {m["MAE"]:.2f} ± {m["STD"]:.2f}')
    ax.text(0.05, 0.95, txt, transform=ax.transAxes,
            va='top', fontsize=10, bbox=dict(boxstyle='round', fc='white', ec='black'))
    ax.legend(loc='lower right'); ax.grid(True); ax.set_aspect('equal')
    fig.tight_layout()
    fig.savefig(output_dir / 'predictions_scatter.png', dpi=150)
    fig.savefig(output_dir / 'predictions_scatter.svg')
    plt.close(fig)


def _write_report(best_params, best_score, m, history, output_dir, gt_counts):
    from datetime import datetime
    with open(output_dir / 'optimization_report.txt', 'w') as f:
        f.write('=' * 80 + '\n')
        f.write('                      GA OPTIMISATION REPORT\n')
        f.write('=' * 80 + '\n')
        f.write(f'Date: {datetime.now():%Y-%m-%d %H:%M:%S}\n\n')
        f.write(f'Training samples : {len(gt_counts)}\n')
        f.write(f'Count range      : [{gt_counts.min()} – {gt_counts.max()}] cysts\n')
        f.write(f'Mean ± SD        : {gt_counts.mean():.1f} ± {gt_counts.std():.1f} cysts\n\n')
        f.write('Optimised parameters:\n')
        for name, val in zip(PARAM_NAMES, best_params):
            f.write(f'  {name:<22} : {val:14.6f}\n')
        f.write('\nTraining-set metrics:\n')
        f.write(f'  MAPE : {best_score:8.4f} %\n')
        f.write(f'  MPE  : {m["MPE"]:8.2f} %\n')
        f.write(f'  MAE  : {m["MAE"]:8.2f} ± {m["STD"]:.2f} cysts\n')
        f.write(f'  RMSE : {m["RMSE"]:8.2f} cysts\n')
        f.write(f'  R²   : {m["R_squared"]:8.4f}\n\n')
        f.write(f'  {"Sample":>8}  {"Ground truth":>12}  {"Predicted":>12}  {"Error":>8}  {"% Error":>9}\n')
        f.write('  ' + '-' * 58 + '\n')
        for i in range(len(gt_counts)):
            err = m['predictions'][i] - gt_counts[i]
            pct = abs(err) / gt_counts[i] * 100
            f.write(f'  {i+1:>8}  {int(gt_counts[i]):>12}  {int(m["predictions"][i]):>12}  '
                    f'{int(err):>+8}  {pct:>8.2f}%\n')
        f.write('\n' + '=' * 80 + '\n')


if __name__ == '__main__':
    run_optimization()

File: matlab/run_optimization.m
% run_optimization.m
% Genetic algorithm optimisation of cyst-counting parameters.
%
% Reads the Optimization set from data/, runs the GA, and writes results
% to results/optimization/.
%
% Required MATLAB toolboxes:
%   Image Processing, Statistics & Machine Learning,
%   Parallel Computing, Global Optimization

clear
clc
delete(gcp('nocreate'));

%% Paths
scriptDir  = fileparts(mfilename('fullpath'));
addpath(scriptDir);                         % makes detectCystPeaks visible to parfor workers

dataDir    = fullfile(scriptDir, '..', 'data', '2_Cyst_counting_pipeline', 'Optimization set');
maskDir    = fullfile(dataDir, 'Cyst masks');
outputDir  = fullfile(scriptDir, '..', 'results', 'optimization');

if ~exist(outputDir, 'dir'), mkdir(outputDir); end

%% Load ground truth from Excel
xlsFile = fullfile(dataDir, 'O1_count.xlsx');
if ~isfile(xlsFile)
    error('Ground-truth file not found:\n  %s\nPlace the Zenodo data inside data/.', xlsFile);
end

T = readtable(xlsFile);                     % columns: Sample, "O1 count"
gt_masks = cell(height(T), 1);
gt_am    = zeros(1, height(T));

for i = 1:height(T)
    sampleID     = T{i, 1}{1};             % e.g. 'K01'
    gt_am(i)     = T{i, 2};
    gt_masks{i}  = fullfile(maskDir, sampleIDtoFilename(sampleID));
    if ~isfile(gt_masks{i})
        error('Mask not found for sample %s:\n  %s', sampleID, gt_masks{i});
    end
end

%% Summary
fprintf('\n==================================================\n');
fprintf('  GENETIC ALGORITHM OPTIMISATION\n');
fprintf('==================================================\n\n');
fprintf('Training samples : %d\n', length(gt_am));
fprintf('Ground truth     : [%d – %d] cysts  (mean %.1f ± %.1f)\n\n', ...
    min(gt_am), max(gt_am), mean(gt_am), std(gt_am));

%% Run GA
[bestParams, bestScore, history] = runGA(gt_masks, gt_am, outputDir);

%% Save outputs
save(fullfile(outputDir, 'optimization_results.mat'), ...
     'bestParams', 'bestScore', 'history', 'gt_am', 'gt_masks');

writematrix(bestParams, fullfile(outputDir, 'BestParams.txt'));

if ~isempty(history.generation)
    fid = fopen(fullfile(outputDir, 'FitnessHistory.txt'), 'w');
    fprintf(fid, 'Generation\tBestScore\tMPE\tMAE\tR2\tPopMean\n');
    fclose(fid);
    fitData = [history.generation(:), history.best_score(:), ...
               history.best_MPE(:), history.best_MAE(:), ...
               history.best_R2(:),  history.mean_score(:)];
    writematrix(fitData, fullfile(outputDir, 'FitnessHistory.txt'), ...
                'WriteMode', 'append', 'Delimiter', 'tab');
end

fprintf('\n==================================================\n');
fprintf('  OPTIMISATION COMPLETE\n');
fprintf('==================================================\n');
fprintf('Results written to: %s\n\n', outputDir);

%% -----------------------------------------------------------------------
%  Internal functions
%% -----------------------------------------------------------------------

function [bestParams, bestScore, history] = runGA(gt_masks, gt_am, outputDir)

    numParams = 6;
    lb = [1,  10,  10000, 1,  0.1, 0.01];
    ub = [15, 100, 150000, 15, 3,  1.5];

    % Temporary file for per-generation history
    histFile = fullfile(outputDir, 'ga_history_temp.mat');
    hs = struct('generation', [], 'best_score', [], 'best_MAPE', [], ...
                'best_MPE', [], 'best_MAE', [], 'best_STD', [], ...
                'best_R2', [], 'best_params', [], 'mean_score', [], ...
                'stall_gen', 0, 'best_score_ever', inf);
    save(histFile, 'hs');

    outputFcn = @(options, state, flag) gaOutputFcn(options, state, flag, ...
                                                    gt_masks, gt_am, outputDir);

    options = optimoptions('ga', ...
        'SelectionFcn',       {@selectiontournament, 3}, ...
        'MutationFcn',        {@mutationadaptfeasible, 0.95}, ...
        'EliteCount',         2, ...
        'PopulationSize',     50, ...
        'MaxGenerations',     200, ...
        'CrossoverFraction',  0.5, ...
        'MaxStallGenerations',10, ...
        'FunctionTolerance',  1e-4, ...
        'MigrationFraction',  0.15, ...
        'UseParallel',        false, ...
        'Display',            'off', ...
        'OutputFcn',          outputFcn);

    fitFcn = @(p) fitnessFunctionWrapper(p, gt_masks, gt_am);

    fprintf('%-8s | %-15s | %-10s | %-14s | %-10s | %-10s | %-8s\n', ...
            'Gen', 'Best MAPE (%)', 'MPE (%)', 'MAE ± STD', 'R²', 'Pop Mean', 'Stall');
    fprintf('%s\n', repmat('-', 1, 95));

    [bestParams, bestScore] = ga(fitFcn, numParams, [], [], [], [], lb, ub, [], options);

    % Retrieve history
    if isfile(histFile)
        load(histFile, 'hs');
        history = hs;
        delete(histFile);
        fprintf('\nHistory: %d generations recorded\n', length(history.generation));
    else
        warning('History file missing – convergence plots will be empty.');
        history = struct('generation', []);
    end

    % Final evaluation
    [~, finalMetrics] = fitness_function(bestParams, gt_masks, gt_am);

    % Console report
    fprintf('\n%s\n', repmat('=', 1, 80));
    fprintf('FINAL RESULTS\n');
    fprintf('%s\n', repmat('=', 1, 80));
    paramNames = {'minPeakSepSmall', 'minPeakSepLarge', 'largeCystVolThresh', ...
                  'maxPeakDistMerge', 'thresholdFraction', 'gaussianSigma'};
    fprintf('\nOptimised parameters:\n');
    for i = 1:6
        fprintf('  %-22s : %10.4f\n', paramNames{i}, bestParams(i));
    end
    fprintf('\nTraining-set metrics (n=%d):\n', length(gt_am));
    fprintf('  MAPE : %8.4f %%\n',  bestScore);
    fprintf('  MPE  : %8.2f %%\n',  finalMetrics.MPE);
    fprintf('  MAE  : %8.2f ± %.2f cysts\n', finalMetrics.MAE, finalMetrics.STD);
    fprintf('  RMSE : %8.2f cysts\n', finalMetrics.RMSE);
    fprintf('  R²   : %8.4f\n',     finalMetrics.R_squared);

    fprintf('\n  Sample | Ground truth | Predicted | Error\n');
    fprintf('  %s\n', repmat('-', 1, 46));
    for i = 1:length(gt_am)
        fprintf('  %6d | %12d | %9d | %+5d\n', i, ...
                finalMetrics.ground_truth(i), finalMetrics.predictions(i), ...
                finalMetrics.predictions(i) - finalMetrics.ground_truth(i));
    end

    % Plots and text report
    plotConvergence(history, outputDir);
    plotPredictions(finalMetrics, outputDir);
    writeTextReport(bestParams, bestScore, finalMetrics, history, outputDir, gt_am);
end

% ---- Fitness wrapper ----
function score = fitnessFunctionWrapper(params, gt_masks, gt_am)
    try
        if any(~isfinite(params)) || any(params <= 0)
            score = 1e6; return;
        end
        score = fitness_function(params, gt_masks, gt_am);
        if ~isfinite(score) || score < 0
            score = 1e6;
        end
    catch
        score = 1e6;
    end
end

% ---- GA output function (tracks history, injects diversity on stall) ----
function [state, options, optchanged] = gaOutputFcn(options, state, flag, ...
                                                    gt_masks, gt_am, outputDir)
    persistent lastGen;
    optchanged = false;

    if isempty(lastGen), lastGen = -1; end
    if ~strcmp(flag, 'iter') || state.Generation == lastGen, return; end

    try
        [bestScore, bestIdx] = min(state.Score);
        if ~isfinite(bestScore) || bestScore > 1e5
            lastGen = state.Generation; return;
        end

        bestP = state.Population(bestIdx, :);
        [~, m] = fitness_function(bestP, gt_masks, gt_am);

        if ~isscalar(m.MAPE) || ~isfinite(m.MAPE), lastGen = state.Generation; return; end

        validScores = state.Score(isfinite(state.Score) & state.Score < 1e5);
        meanScore   = mean(validScores);

        histFile = fullfile(outputDir, 'ga_history_temp.mat');
        load(histFile, 'hs');

        if bestScore < hs.best_score_ever - 1e-6
            hs.best_score_ever = bestScore;
            hs.stall_gen = 0;
            fprintf('%-8d | %15.4f | %10.2f | %6.2f±%-4.2f | %10.4f | %10.4f\n', ...
                    state.Generation, bestScore, m.MPE, m.MAE, m.STD, m.R_squared, meanScore);
        else
            hs.stall_gen = hs.stall_gen + 1;
            fprintf('%-8d | %15.4f | %10.2f | %6.2f±%-4.2f | %10.4f | %10.4f | %8d\n', ...
                    state.Generation, bestScore, m.MPE, m.MAE, m.STD, m.R_squared, meanScore, hs.stall_gen);
        end

        % Diversity injection after 5 stall generations
        if hs.stall_gen >= 5
            fprintf('   >>> Diversity injection: reinitialising 50%% of population <<<\n');
            lb       = [1, 10, 10000, 1, 0.1, 0.01];
            ub       = [15, 100, 150000, 15, 3, 1.5];
            popSize  = size(state.Population, 1);
            nReinit  = round(0.5 * (popSize - 2));  % keep 2 elite
            idxReinit = randperm(popSize - 2, nReinit) + 2;
            for k = idxReinit
                state.Population(k, :) = lb + rand(1, 6) .* (ub - lb);
            end
            hs.stall_gen = 0;
            optchanged   = true;
        end

        hs.generation(end+1)   = state.Generation;
        hs.best_score(end+1)   = bestScore;
        hs.best_MPE(end+1)     = m.MPE;
        hs.best_MAE(end+1)     = m.MAE;
        hs.best_STD(end+1)     = m.STD;
        hs.best_R2(end+1)      = m.R_squared;
        hs.mean_score(end+1)   = meanScore;
        if isempty(hs.best_params)
            hs.best_params = bestP(:)';
        else
            hs.best_params(end+1,:) = bestP(:)';
        end

        save(histFile, 'hs');
        lastGen = state.Generation;

    catch ME
        fprintf('Warning (generation %d): %s\n', state.Generation, ME.message);
        lastGen = state.Generation;
    end
end

% ---- Convergence plots ----
function plotConvergence(history, outputDir)
    if isempty(history.generation), return; end

    fig = figure('Position', [100, 100, 1400, 900], 'Visible', 'off');

    subplot(2,3,1)
    plot(history.generation, history.best_score, 'b-o', 'LineWidth', 2, 'MarkerSize', 5);
    hold on;
    plot(history.generation, history.mean_score, 'r--', 'LineWidth', 1.5);
    xlabel('Generation'); ylabel('MAPE (%)');
    title('Fitness convergence'); legend('Best', 'Population mean', 'Location', 'best');
    grid on;

    subplot(2,3,2)
    errorbar(history.generation, history.best_MAE, history.best_STD, ...
             'b-o', 'LineWidth', 2, 'MarkerSize', 5, 'CapSize', 8);
    xlabel('Generation'); ylabel('MAE (cysts)');
    title('MAE ± STD'); grid on;

    subplot(2,3,3)
    plot(history.generation, history.best_R2, 'g-o', 'LineWidth', 2, 'MarkerSize', 5);
    xlabel('Generation'); ylabel('R²');
    title('R² convergence'); ylim([0, 1]); grid on;

    subplot(2,3,4)
    plot(history.generation, history.best_MPE, 'm-o', 'LineWidth', 2, 'MarkerSize', 5);
    xlabel('Generation'); ylabel('MPE (%)');
    title('Bias convergence'); yline(0, 'k--', 'LineWidth', 1.5); grid on;

    subplot(2,3,[5 6])
    lb = [1, 10, 10000, 1, 0.1, 0.01];
    ub = [15, 100, 150000, 15, 3, 1.5];
    pNorm = (history.best_params - lb) ./ (ub - lb);
    pLabels = {'P1:minPeakSmall','P2:minPeakLarge','P3:volThresh', ...
               'P4:maxDistMerge','P5:threshFrac','P6:gaussSigma'};
    plot(history.generation, pNorm, '-o', 'LineWidth', 1.5, 'MarkerSize', 4);
    xlabel('Generation'); ylabel('Normalised value [0–1]');
    title('Parameter evolution');
    legend(pLabels, 'Location', 'eastoutside', 'FontSize', 9);
    grid on;

    sgtitle('GA Convergence', 'FontSize', 14, 'FontWeight', 'bold');
    saveas(fig, fullfile(outputDir, 'GA_convergence.png'));
    saveas(fig, fullfile(outputDir, 'GA_convergence.fig'));
    close(fig);
    fprintf('Convergence plot saved.\n');
end

% ---- Prediction scatter plot ----
function plotPredictions(m, outputDir)
    fig = figure('Position', [100, 100, 700, 650], 'Visible', 'off');

    gt   = m.ground_truth(:)';
    pred = m.predictions(:)';
    rng  = [min([gt, pred]), max([gt, pred])];
    margin = diff(rng) * 0.1;
    ax   = [rng(1)-margin, rng(2)+margin];

    scatter(gt, pred, 100, 'filled', 'MarkerFaceAlpha', 0.7, 'MarkerFaceColor', [0.2 0.4 0.8]);
    hold on;
    plot(ax, ax, 'k--', 'LineWidth', 2);
    p = polyfit(gt, pred, 1);
    xf = linspace(ax(1), ax(2), 100);
    plot(xf, polyval(p, xf), 'r-', 'LineWidth', 2);

    xlabel('Ground truth (cysts)', 'FontWeight', 'bold');
    ylabel('Predicted (cysts)',    'FontWeight', 'bold');
    title('Predictions vs ground truth (training set)', 'FontWeight', 'bold');

    R2 = m.R_squared; if numel(R2) > 1, R2 = R2(1); end
    annotation_str = sprintf('y = %.3fx + %.2f\nMAPE: %.2f%%\nR²: %.4f\nMAE: %.2f ± %.2f', ...
                             p(1), p(2), m.MAPE, R2, m.MAE, m.STD);
    xl = xlim; yl = ylim;
    text(xl(1)+0.05*diff(xl), yl(2)-0.05*diff(yl), annotation_str, ...
         'VerticalAlignment', 'top', 'BackgroundColor', 'w', ...
         'EdgeColor', 'k', 'FontSize', 10);

    legend({'Samples', 'Identity', 'Regression'}, 'Location', 'southeast');
    grid on; axis square; xlim(ax); ylim(ax);
    set(gca, 'FontSize', 11, 'LineWidth', 1.5); box on;

    saveas(fig, fullfile(outputDir, 'predictions_scatter.png'));
    saveas(fig, fullfile(outputDir, 'predictions_scatter.fig'));
    saveas(fig, fullfile(outputDir, 'predictions_scatter.svg'));
    close(fig);
    fprintf('Prediction scatter plot saved.\n');
end

% ---- Text report ----
function writeTextReport(bestParams, bestScore, m, history, outputDir, gt_am)
    fid = fopen(fullfile(outputDir, 'optimization_report.txt'), 'w');
    fprintf(fid, '================================================================================\n');
    fprintf(fid, '                        GA OPTIMISATION REPORT\n');
    fprintf(fid, '================================================================================\n');
    fprintf(fid, 'Date: %s\n\n', datestr(now, 'yyyy-mm-dd HH:MM:SS'));

    fprintf(fid, '--- Dataset ---\n');
    fprintf(fid, 'Training samples : %d\n', length(gt_am));
    fprintf(fid, 'Count range      : [%d – %d] cysts\n', min(gt_am), max(gt_am));
    fprintf(fid, 'Mean ± SD        : %.1f ± %.1f cysts\n\n', mean(gt_am), std(gt_am));

    fprintf(fid, '--- Optimised parameters ---\n');
    pNames = {'minPeakSepSmall','minPeakSepLarge','largeCystVolThresh', ...
              'maxPeakDistMerge','thresholdFraction','gaussianSigma'};
    for i = 1:6
        fprintf(fid, '  %-22s : %12.6f\n', pNames{i}, bestParams(i));
    end

    fprintf(fid, '\n--- Training-set metrics ---\n');
    fprintf(fid, '  MAPE : %8.4f %%\n', bestScore);
    fprintf(fid, '  MPE  : %8.2f %%\n', m.MPE);
    fprintf(fid, '  MAE  : %8.2f ± %.2f cysts\n', m.MAE, m.STD);
    fprintf(fid, '  RMSE : %8.2f cysts\n', m.RMSE);
    fprintf(fid, '  R²   : %8.4f\n\n', m.R_squared(1));

    fprintf(fid, '--- Per-sample predictions ---\n');
    fprintf(fid, '  %-8s | %-12s | %-12s | %-10s | %-10s\n', ...
            'Sample', 'Ground truth', 'Predicted', 'Error', '%% Error');
    fprintf(fid, '  %s\n', repmat('-', 1, 62));
    for i = 1:length(gt_am)
        pctErr = 100 * abs(m.predictions(i) - gt_am(i)) / gt_am(i);
        fprintf(fid, '  %-8d | %12d | %12d | %+10d | %9.2f%%\n', ...
                i, gt_am(i), m.predictions(i), m.predictions(i) - gt_am(i), pctErr);
    end

    fprintf(fid, '\n================================================================================\n');
    fclose(fid);
    fprintf('Text report saved.\n');
end

% ---- Sample ID to NIfTI filename ----
function fname = sampleIDtoFilename(id)
    if id(1) == 'K'
        prefix = 'KRATS_';
    else
        prefix = 'XRATS_';
    end
    fname = sprintf('%s%03d.nii.gz', prefix, str2double(id(2:end)));
end
